File: CSES/8. String Algorithms/CSES-1731_Word Combinations.cpp
#include <bits/stdc++.h>
using namespace std;
using LL = long long;
const LL MOD = 1e9 + 7;
#define endl '\n'

class TrieNode {
public:
    TrieNode* child[26];
    bool is_end;
    TrieNode() {
        memset(child, 0, sizeof(child));
        is_end = false;
    }
};

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    int n, k;
    string s, word;
    cin >> s;
    n = s.size();
    cin >> k;
    TrieNode* root = new TrieNode();
    while (k--) {
        cin >> word;
        TrieNode* node = root;
        for (char ch : word) {
            int idx = ch - 'a';
            if (node->child[idx] == nullptr) {
                node->child[idx] = new TrieNode();
            }
            node = node->child[idx];
        }
        node->is_end = true;
    }
    vector<LL> f(n + 1);
    f[n] = 1;
    for (int i = n - 1; i >= 0; --i) {
        TrieNode* node = root;
        for (int j = i; j < n; ++j) {
            int idx = s[j] - 'a';
            if (node->child[idx] == nullptr) break;
            node = node->child[idx];
            if (node->is_end) {
                f[i] = (f[i] + f[j + 1]) % MOD;
            }
        }
    }
    cout << f[0] << endl;
    return 0;
}

File: CSES/8. String Algorithms/CSES-1731_Word Combinations.py
"""
    Trie DP

    Python TLE
"""
import sys
input = sys.stdin.readline
def print(val=""): sys.stdout.write(str(val) + "\n")

class TrieNode:
    def __init__(self):
        self.child = [None] * 26
        self.is_end = False

MOD = 10 ** 9 + 7
s = input().strip()
n = len(s)
root = TrieNode()
k = int(input())
for _ in range(k):
    w = input().strip()
    node = root
    for ch in w:
        idx = ord(ch) - ord('a')
        if node.child[idx] is None:
            node.child[idx] = TrieNode()
        node = node.child[idx]
    node.is_end = True

f = [0] * (n + 1)
f[n] = 1
for i in range(n - 1, -1, -1):
    node = root
    for j in range(i, n):
        idx = ord(s[j]) - ord('a')
        if node.child[idx] is None:
            break
        node = node.child[idx]
        if node.is_end:
            f[i] = (f[i] + f[j + 1]) % MOD
print(f[0])
